File: app/src/main/java/moe/notify/animenotifier/storage/AnimeListRepositoryImpl.java
package moe.notify.animenotifier.storage;

import android.content.Context;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import moe.notify.animenotifier.domain.model.animelist.AnimeList;
import moe.notify.animenotifier.domain.repository.AnimeListRepository;
import moe.notify.animenotifier.network.RestClient;
import moe.notify.animenotifier.network.converters.RESTAnimeListModelConverter;
import moe.notify.animenotifier.network.model.animelist.RESTAnimeList;
import moe.notify.animenotifier.network.services.SyncService;
import moe.notify.animenotifier.storage.converters.StorageAnimeListModelConverter;
import retrofit2.Response;
import timber.log.Timber;


public class AnimeListRepositoryImpl implements AnimeListRepository {

    private final Context context;

    public AnimeListRepositoryImpl(Context context) {
        this.context = context;
    }

    public AnimeListRepositoryImpl() {
        context = null;
    }

    @Override
    public void insert(AnimeList item) {
        moe.notify.animenotifier.storage.model.animelist.AnimeList dbItem = new moe.notify.animenotifier.storage.model.animelist.AnimeList();
        StorageAnimeListModelConverter.convertAnimeListToStorageModel(item, dbItem);

        dbItem.insert();
    }

    @Override
    public void update(AnimeList item) {
        moe.notify.animenotifier.storage.model.animelist.AnimeList dbItem = new moe.notify.animenotifier.storage.model.animelist.AnimeList();
        StorageAnimeListModelConverter.convertAnimeListToStorageModel(item, dbItem);

        dbItem.update();
    }

    @Override
    public AnimeList getByUser(String user) {
        moe.notify.animenotifier.storage.model.animelist.AnimeList savedAnimeList = /*SQLite
                .select()
                .from(moe.notify.animenotifier.storage.model.animelist.AnimeList.class)
                .where(AnimeList_Watching_Table.animeList_user.is(user))
                .querySingle();*/null;

        AnimeList animeList = new AnimeList();
        if (savedAnimeList == null) {
            savedAnimeList = new moe.notify.animenotifier.storage.model.animelist.AnimeList();
            SyncService syncService = RestClient.getService(SyncService.class);
            try {
                Response<RESTAnimeList> response = syncService.getAnimeListFromUser("Scott").execute();

                RESTAnimeListModelConverter.convertToDomainModel(response.body(), animeList);

                // everything went well, lets update
                if (response.isSuccessful()) {
                    StorageAnimeListModelConverter.convertAnimeListToStorageModel(animeList, savedAnimeList);

//                mAnimeRepository.markSynced(mUnsyncedAnimes);
                }

            } catch (IOException e) {
                Timber.e(e, "A problem occur while trying to talk with the server");
            }
        }
        StorageAnimeListModelConverter.convertAnimeListToDomainModel(savedAnimeList, animeList);

        return animeList;
    }

    @Override
    public List<AnimeList> getAll() {
//        List<moe.notify.animenotifier.storage.model.animelist.AnimeList> savedAnimeList = SQLite
//                .select()
//                .from(moe.notify.animenotifier.storage.model.animelist.AnimeList.class)
//                .queryList();
//        List<AnimeList> animeList = new ArrayList<>();
//
//        StorageAnimeListModelConverter.convertAnimeListToDomainModel(savedAnimeList, animeList);
        return null;
    }

    @Override
    public List<AnimeList> getAllUnsynced() {
        // When we will be able to push data to the server.
        return new ArrayList<>();
    }

    @Override
    public void markSynced(List<AnimeList> animeLists) {
        // When we will be able to push data to the server.
    }

    @Override
    public void delete(AnimeList animeList) {

    }
}

File: app/src/main/java/moe/notify/animenotifier/storage/AnimeListRepositoryImpl.kt
package moe.notify.animenotifier.storage

import android.content.Context
import io.realm.Realm
import io.realm.RealmList
import moe.notify.animenotifier.domain.model.animelist.AnimeList
import moe.notify.animenotifier.domain.repository.AnimeListRepository
import moe.notify.animenotifier.network.RestClient
import moe.notify.animenotifier.network.services.SyncService
import moe.notify.animenotifier.sync.SyncAdapter
import timber.log.Timber
import java.io.IOException
import java.util.ArrayList

open class AnimeListRepositoryImpl : AnimeListRepository {

  private val context: Context?

  constructor(context: Context) {
    this.context = context
  }

  constructor() {
    context = null
  }

  override fun insert(animeList: AnimeList) {
    Realm.getDefaultInstance().use { realmInstance ->
      realmInstance.executeTransaction { realm ->
        realm.insert(animeList)
      }
    }
  }

  override fun update(animeList: AnimeList) {
    Realm.getDefaultInstance().use { realmInstance ->
      realmInstance.executeTransaction { realm ->
        realm.insertOrUpdate(animeList)
      }
    }
  }

  override fun getByUser(user: String): AnimeList {
    var realm: Realm? = null
    var animeList: AnimeList = AnimeList()
    try {
      realm = Realm.getDefaultInstance()
      val savedAnimeList = realm.where(AnimeList::class.java).equalTo("user", user).findFirst()
      if (savedAnimeList == null) {
        val syncService = RestClient.getService(SyncService::class.java)
        try {
          val response = syncService.getAnimeListFromUser("Scott").execute()
          if (response.isSuccessful) {
            animeList = response.body()
            realm.executeTransaction { realm ->
              realm.insert(animeList)
            }
          }
        } catch (e: IOException) {
          Timber.e(e, "A problem occur while trying get the animeList of the user:%l,", user)
        }

      } else {
        // TODO 27/12/2016 Return a default Anime
//          StorageAnimeModelConverter.convertToDomainModel(savedAnimeList, anime);
        val syncService = RestClient.getService(SyncService::class.java)
        try {
          val response = syncService.getAnimeListFromUser("Scott").execute()
          if (response.isSuccessful) {
            animeList = response.body()
            realm.executeTransactionAsync { realm ->
              realm.insert(animeList)
            }
          }
        } catch (e: IOException) {
          Timber.e(e, "A problem occur while trying get the animeList of the user:%l,", user)
        }
        animeList = realm.copyFromRealm(savedAnimeList)
      }
    } catch (e: Exception) {
      Timber.e(e, "meh")
    } finally {
      if (realm != null) {
        realm.close()
      }
    }
    return animeList
  }

  override fun getAll(): RealmList<AnimeList>? {
    val animes: RealmList<AnimeList> = RealmList()
    Realm.getDefaultInstance().use { realmInstance ->
      val savedAnimes = realmInstance.where(AnimeList::class.java).findAll()
      if (savedAnimes != null) {
        animes.addAll(savedAnimes)
      }
      return animes
    }
  }

  override fun getAllUnsynced(): List<AnimeList> {
    // When we will be able to push data to the server.
    return ArrayList()
  }

  override fun markSynced(animeLists: List<AnimeList>) {
    // When we will be able to push data to the server.
  }

  override fun delete(animeList: AnimeList) {
    Realm.getDefaultInstance().use { realmInstance ->
      val savedAnimes = realmInstance.where(AnimeList::class.java).equalTo("user",
          animeList.user).findFirst()
      if (savedAnimes != null) {
        realmInstance.executeTransaction { realm ->
          animeList.deleteFromRealm()
        }
      }
    }
    SyncAdapter.triggerSync(context)
  }
}

